File: src/main/java/com/szucsatti/lemkehowson/Matrix.java
package com.szucsatti.lemkehowson;

import lombok.extern.java.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.szucsatti.lemkehowson.Rational.*;


@Log
public class Matrix {

    private static final String START_LINE = "|  ";
    private static final String END_LINE = "  |";
    private static final String VALUE_DELIMITER = "  |  ";
    private Rational[][] matrix;
    private int rows;
    private int cols;

    public static final Matrix IDENTITY_2X2 = new Matrix(new Rational[][]{
            {ONE, ZERO},
            {ZERO, ONE}});

    public static final Matrix COLUMN_ONE = new Matrix(new Rational[][]{
            {ONE},
            {ONE}});

    public Matrix(final Rational[][] matrix) {
        this.matrix = matrix;
        this.rows = matrix.length;
        this.cols = matrix[0].length;
    }

    public Matrix(final long[][] matrix) {
        this.rows = matrix.length;
        this.cols = matrix[0].length;
        this.matrix = new Rational[this.rows][this.cols];
        for(int row = 0; row < this.getRows(); row++){
            for(int col = 0; col < this.getCols(); col++){
                this.matrix[row][col] = valueOf(matrix[row][col]);
            }
        }
    }

    public Matrix(final Matrix other){
        this.rows = other.getRows();
        this.cols = other.getCols();
        this.matrix = new Rational[this.getRows()][this.getCols()];

        for(int rowIndex = 0; rowIndex < rows; rowIndex++){
            Rational[] row = other.matrix[rowIndex];
            System.arraycopy(row, 0, this.matrix[rowIndex], 0, row.length);
        }
    }

    public Matrix join(final Matrix other){
        assert this.getRows() == other.getRows();
        Rational[][] joined = new Rational[this.getRows()][this.getCols() + other.getCols()];

        for(int row = 0; row < other.getRows(); row++){
            Rational[] otherRow = other.matrix[row];
            Rational[] thisRow = this.matrix[row];
            System.arraycopy(thisRow, 0, joined[row], 0, thisRow.length);
            System.arraycopy(otherRow, 0, joined[row], thisRow.length, otherRow.length);
        }

        return new Matrix(joined);
    }

    public Matrix[] split(){
        int halfCol = this.getCols() / 2;

        Rational [][] firstMatrix = new Rational[this.getRows()][halfCol];
        Rational [][] secondMatrix = new Rational[this.getRows()][halfCol];

        for(int row = 0; row < rows; row++){
            Rational[] thisRow = this.matrix[row];
            System.arraycopy(thisRow, 0, firstMatrix[row], 0, halfCol);
            System.arraycopy(thisRow, halfCol, secondMatrix[row], 0, halfCol) ;
        }

        return new Matrix[]{new Matrix(firstMatrix), new Matrix(secondMatrix)};
    }

    public Rational get(int row, int column) {
        return matrix[row][column];
    }

    public Matrix normalize() {
        Rational normalizationConstant = getNormalizationConstant();
        Rational[][] normalized = new Rational[this.getRows()][this.getCols()];
        if(normalizationConstant.isGreaterOrEqualThan(ZERO)){
            for(int row = 0; row < rows; row++){
                for(int col = 0; col < cols; col++){
                    normalized[row][col] = this.matrix[row][col].plus(normalizationConstant);
                }
            }
        }
        return new Matrix(normalized);
    }

    private Rational getNormalizationConstant(){
        Rational minimum = getMinimumValue();
        return minimum.isGreaterOrEqualThan(ZERO) ? ZERO : minimum.absValue().plus(ONE);
    }

    protected Rational getMinimumValue() {
        Rational minimum = matrix[0][0];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (matrix[row][col].isLessThan(minimum)) {
                    minimum = matrix[row][col];
                }
            }
        }
        return minimum;
    }

    public void print() {
        for (int row = 0; row < rows; row++) {
            System.out.print(START_LINE);
            for (int col = 0; col < cols; col++) {
                System.out.print(matrix[row][col]);
                if (col != cols - 1) {
                    System.out.print(VALUE_DELIMITER);
                }
            }
            System.out.print(END_LINE);
            if (row != rows - 1) {
                System.out.println();
            }
        }
        System.out.println();
    }

    public int getCols(){
        return cols;
    }

    public int getRows(){
        return rows;
    }

    public Rational ratio(int uRow, int uCol, int lRow, int lCol){
        return matrix[uRow][uCol].divide(matrix[lRow][lCol]);
    }

    public Matrix multiplyRow(final int rowIndex, final Rational multiplyBy){
        Rational[][] multipliedMatrix = this.matrix;

        for(int col = 0; col < getCols(); col++){
            multipliedMatrix[rowIndex][col] = multipliedMatrix[rowIndex][col].times(multiplyBy);
        }

        return new Matrix(multipliedMatrix);
    }

    public Matrix subtract(final int rowIndex, final int fromIndex){
        Rational[][] subtractedMatrix = this.matrix;

        for(int col = 0; col < this.getCols(); col++){
            subtractedMatrix[fromIndex][col] = subtractedMatrix[fromIndex][col].minus(subtractedMatrix[rowIndex][col]);
        }
        return new Matrix(subtractedMatrix);
    }

    public List<Integer> getNoZeroCols(){
        List<Integer> nonZeroCols = new ArrayList<>();
        for(int col = 0; col < getCols(); col++){
            boolean nonZero = true;
            for(int row = 0; row < getRows(); row++){
                if(Rational.ZERO.equals(matrix[row][col])){
                    nonZero = false;
                    break;
                }
            }
            if(nonZero){
                nonZeroCols.add(col);
            }
        }
        return nonZeroCols;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Matrix matrix1 = (Matrix) o;

        if (rows != matrix1.rows) return false;
        if (cols != matrix1.cols) return false;
        return Arrays.deepEquals(matrix, matrix1.matrix);
    }

    @Override
    public int hashCode() {
        int result = Arrays.deepHashCode(matrix);
        result = 31 * result + rows;
        result = 31 * result + cols;
        return result;
    }
}

File: src/main/kotlin/com/szucsatti/lemkehowsonkt/Matrix.kt
package com.szucsatti.lemkehowsonkt

import org.jscience.mathematics.number.Rational
import org.jscience.mathematics.number.Rational.ONE
import org.jscience.mathematics.number.Rational.ZERO


internal typealias MatrixOfRationals = Array<Array<Rational>>
internal typealias MatrixOfAny = Array<Array<Any>>

fun MatrixOfAny.rows() = size
fun MatrixOfAny.cols() = this[0].size

fun MatrixOfRationals.rows() = size
fun MatrixOfRationals.cols() = this[0].size

fun Array<Rational>.min(): Rational {
    return reduce { left: Rational, right: Rational -> if (left.isLessThan(right)) left else right }
}

class Matrix(private val matrix: MatrixOfRationals) {

    private val rows: Int = matrix.rows()
    private val cols: Int = matrix.cols()

    /* ==================================================================== */
    /* Operations on this matrix                                            */
    /* ==================================================================== */

    fun copy(): Matrix {
        var result = init(rows, cols)

        matrix.forEachIndexed { index, row -> row.copyInto(result[index]) }

        return Matrix(result)
    }

    fun min(): Rational {
        return matrix.asSequence()
                .flatMap { sequenceOf(it.min()) }
                .reduce { left, right -> if (left.isLessThan(right)) left else right }
    }

    fun print() {
        for (row in 0 until rows) {
            print(START_LINE)
            for (col in 0 until cols) {
                print(matrix[row][col].toText())
                if (col != cols - 1) {
                    print(VALUE_DELIMITER)
                }
            }
            print(END_LINE)
            if (row != rows - 1) {
                println()
            }
        }
        println()
    }

    fun normalize(): Matrix {
        val normalizationConstant: Rational = normalizationConstant()
        val normalized = init(rows, cols)
        if (normalizationConstant.isGreaterOrEqualThan(ZERO)) {
            matrix.forEachIndexed { i, row ->
                run {
                    row.forEachIndexed { j, _ -> normalized[i][j] = matrix[i][j].plus(normalizationConstant) }
                }
            }
        }
        return Matrix(normalized)
    }

    private fun normalizationConstant(): Rational {
        val min = min()
        return if (min.isGreaterOrEqualThan(ZERO)) ZERO else min.abs().plus(ONE)
    }

    fun subtract(rowIndex: Int, fromIndex: Int): Matrix {
        val subtractedMatrix = copy().matrix

        subtractedMatrix[fromIndex] = subtractedMatrix[fromIndex].mapIndexed { index, rational ->
            rational.minus(subtractedMatrix[rowIndex][index])
        }.toTypedArray()

        return Matrix(subtractedMatrix)
    }

    fun nonZeroCols(): List<Int> {
        val nonZeroCols= mutableListOf<Int>()

        for(col in 0 until cols ) {
            if(matrix.map {it[col]}.contains(ZERO).not())
                nonZeroCols.add(col)
        }

        return nonZeroCols
    }

    fun switchRowsWithCols(): Matrix {
        val result = init(rows, cols)
        for (row in 0 until rows) {
            for (col in 0 until cols) {
                result[col][row] = matrix[row][col]
            }
        }
        return Matrix(result)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val typedOther = other as Matrix
        if (rows != typedOther.rows) return false

        return if (cols != typedOther.cols) false else matrix.contentDeepEquals(typedOther.matrix)
    }

    override fun hashCode(): Int {
        var result = matrix.contentDeepHashCode()
        result = 31 * result + rows
        result = 31 * result + cols
        return result
    }

    /* ==================================================================== */
    /* Operations with other matrices                                       */
    /* ==================================================================== */

    fun join(other: Matrix): Matrix {
        assert(rows == other.rows)
        val result = init(rows, cols + other.cols)

        matrix.forEachIndexed { index, row ->
            run {
                row.copyInto(result[index])
                other.matrix[index].copyInto(result[index], destinationOffset = row.size)
            }
        }

        return Matrix(result)
    }

    fun split(): Array<Matrix> {
        val halfCol = cols / 2

        val firstMatrix = init(rows, halfCol)
        val secondMatrix = init(rows, halfCol)

        matrix.forEachIndexed { index, row ->
            run {
                row.copyInto(firstMatrix[index], endIndex = halfCol)
                row.copyInto(secondMatrix[index], startIndex = halfCol)
            }
        }

        return arrayOf(Matrix(firstMatrix), Matrix(secondMatrix))
    }

    fun multiplyRow(rowIndex: Int, multiplyBy: Rational): Matrix {
        val multipliedMatrix = copy().matrix

        multipliedMatrix[rowIndex] = multipliedMatrix[rowIndex]
                .map { it.times(multiplyBy) }
                .toTypedArray()

        return Matrix(multipliedMatrix)
    }

    companion object {
        private const val START_LINE = "|  "
        private const val END_LINE = "  |"
        private const val VALUE_DELIMITER = "  |  "

        val IDENTITY_2X2 = Matrix(arrayOf(
                arrayOf(ONE, ZERO),
                arrayOf(ONE, ZERO)))
        val COLUMN_ONE = Matrix(arrayOf(
                arrayOf(ONE),
                arrayOf(ONE)
        ))

        private fun init(rows: Int, cols: Int) = Array(rows) { Array<Rational>(cols) { ONE } }

        fun build(vararg values: Array<Any>): Matrix {
            @Suppress("UNCHECKED_CAST")
            return buildFromMatrixOfAny(values as MatrixOfAny)
        }

        private fun buildFromMatrixOfAny(values: MatrixOfAny): Matrix {
            var result = init(values.rows(), values.cols())
            for (row in values.indices) {
                for (col in values[0].indices) {
                    result[row][col] = valueOf(values[row][col])
                }
            }
            return Matrix(result)
        }
    }
}